src/EditorFeatures/Core/ValueTracking/ValueTrackingService.cs

```diff
@@ -23,101 +35,355 @@ public ValueTrackingService()
-            private readonly ValueTrackingProgressCollector _valueTrackingProgressCollector;
-            public FindReferencesProgress(ValueTrackingProgressCollector valueTrackingProgressCollector)
+            var containingMethod = (IMethodSymbol)parameterSymbol.ContainingSymbol;
+            var findReferenceProgressCollector = new FindReferencesProgress(progressCollector);
+            await SymbolFinder.FindReferencesAsync(
+                containingMethod,
+                document.Project.Solution,
+                findReferenceProgressCollector,
+                documents: null, FindReferencesSearchOptions.Default, cancellationToken).ConfigureAwait(false);
+        }
+
+        private static async Task TrackMethodSymbolAsync(IMethodSymbol methodSymbol, Document document, ValueTrackingProgressCollector progressCollector, CancellationToken cancellationToken)
+        {
+            var hasAnyOutData = HasAValueReturn(methodSymbol) || HasAnOutOrRefParam(methodSymbol);
+            if (!hasAnyOutData)
             {
-                _valueTrackingProgressCollector = valueTrackingProgressCollector;
+                // With no out data, there's nothing to do here
+                return;
             }
 
-            public IStreamingProgressTracker ProgressTracker => this;
+            // TODO: Use DFA to find meaningful returns? https://github.com/dotnet/roslyn-analyzers/blob/9e5f533cbafcc5579e4d758bc9bde27b7611ca54/docs/Writing%20dataflow%20analysis%20based%20analyzers.md 
+            if (HasAValueReturn(methodSymbol))
+            {
+                foreach (var location in methodSymbol.GetDefinitionLocationsToShow())
+                {
+                    if (location.SourceTree is null)
+                    {
+                        continue;
+                    }
 
-            public ValueTask AddItemsAsync(int count) => new();
+                    var node = location.FindNode(cancellationToken);
+                    var sourceDoc = document.Project.Solution.GetRequiredDocument(location.SourceTree);
+                    var syntaxFacts = sourceDoc.GetRequiredLanguageService<ISyntaxFactsService>();
+                    var returnStatements = node.DescendantNodesAndSelf().Where(n => syntaxFacts.IsReturnStatement(n)).ToImmutableArray();
+                    var semanticModel = await sourceDoc.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
 
-            public ValueTask ItemCompletedAsync() => new();
+                    if (returnStatements.IsDefaultOrEmpty)
+                    {
+                        // If there are no return statements and the method has a return type, then the method body is an expression
+                        // and we're interested in parsing that expression
+                        var expression = node.DescendantNodesAndSelf().First(syntaxFacts.IsMethodBody);
+                        if (expression is null)
+                        {
+                            return;
```

Comment:
Should this be continue, so processing out params isn't skipped?

Reply:
Let me get a test for out params, but I think you're right.

Reply:
It does need to continue, but there are some nuances that need to be handled for adding tests. I'm going to do that in the next PR and address this issue. Will link once the PR is up